iproto: override request handlers using event triggers

Gumix opened the pull request:
Closes #8138

drewdzzz commented on src/box/iproto.cc:
```diff
@@ -3291,6 +3425,81 @@ iproto_thread_init(struct iproto_thread *iproto_thread)
+		*is_by_id = true;
+		type = strtol(req_name, &end, 10);
+		if (req_name == end)
+			return iproto_type_MAX;
```
`box.iproto.override[select]` or `box.iproto.override[15].header` are invalid names which could very possibly come here - they should be logged too.
More strictly, let's log change on every event, that belongs to `box.iproto.override` namespace but does not override any request handler.

Gumix replied:
Done.

drewdzzz commented on src/box/iproto.cc:
```diff
@@ -2815,9 +2815,16 @@ tx_run_override_triggers(struct event *event, const char *header,
+		func_adapter_push_msgpack(trigger, &ctx, header, header_end,
+					  &mp_ctx_header);
+		func_adapter_push_msgpack(trigger, &ctx, body, body_end,
+					  &mp_ctx_body);
```
Lua hashes are used for translation, while `func_adapter` was designed to call arbitrary functions.
Now `func_adapter` is used only for Lua, but we should be ready for new `func_adapter` implementations.
Probably, it will be a good solution to add a used hash function to `struct mp_ctx` in the future - then other languages will use `lua_hash`, because with high probability only Lua has pre-calculated hash on its values.
Let's discuss it @alyapunov before merging this PR.

drewdzzz commented on test/unit/lua_func_adapter.c:
```diff
@@ -270,7 +270,7 @@ test_msgpack(void)
 	struct func_adapter_ctx ctx;
 	func_adapter_begin(func, &ctx);
-	func_adapter_push_msgpack(func, &ctx, mp_buf, mp);
+	func_adapter_push_msgpack(func, &ctx, mp_buf, mp, NULL);
```
Let's factor `func_adapter` update out to a separate commit and add a unit test for translation.

Gumix replied:
Done.

drewdzzz commented on src/box/space.c:
```diff
@@ -813,12 +813,12 @@ space_run_replace_triggers(struct event *event, struct txn *txn,
 			func_adapter_push_msgpack(trigger, &ctx, row->header,
-						  row->header_end);
+						  row->header_end, NULL);
 			assert(row->bodycnt == 1);
 			const char *body = row->body[0].iov_base;
 			const char *body_end = body + row->body[0].iov_len;
 			func_adapter_push_msgpack(trigger, &ctx, body,
-						  body_end);
+						  body_end, NULL);
```
We could add translation to the space recovery triggers (as a separate commit).

Gumix replied:
Done.

drewdzzz commented on src/box/lua/iproto.lua:
```diff
@@ -0,0 +1,28 @@
+                  'Usage: box.iproto.override(request_type, callback)')
+    end
+    if type(request_type) ~= 'number' then
+        box.error(box.error.PROC_LUA, 'expected uint64_t as 1 argument')
```
There is no `uint64_t` in Lua.

Gumix replied:
Fixed.

drewdzzz commented on src/box/lua/iproto.lua:
```diff
@@ -0,0 +1,28 @@
+    end
+
+    local event_name = ('box.iproto.override[%d]'):format(request_type)
+    local trigger_name = '__tarantool__internal__iproto__override__trigger__'
```
I think it's OK to set a handler with `box.iproto.override` and then delete it with trigger registry.
So let's call a trigger with more readable name (`box.iproto.override.lua.user_handler`, for example).

Gumix replied:
Done.

xuniq commented on changelogs/unreleased/gh-8138-iproto-override-via-event-triggers.md:
```diff
@@ -0,0 +1,10 @@
+## feature/core
+
+* Now it is possible to set IPROTO request handler callbacks by using the new
+  Lua module `trigger`, e.g.:
```
```suggestion
  Lua module `trigger`, for example:
```

Gumix replied:
Done.

xuniq commented on changelogs/unreleased/gh-8138-iproto-override-via-event-triggers.md:
````diff
@@ -0,0 +1,10 @@
+  local trigger = require('trigger')
+  trigger.set('box.iproto.override.select', 'my_select', my_select_handler)
+  ```
+  This new method works before the initial `box.cfg{}` call, also it allows to
+  set multiple handlers for a single request type (gh-8138).
````
```suggestion
  The method works before the initial `box.cfg{}` call. Also, the method allows settinng multiple handlers for a single request type (gh-8138).
```

Gumix replied:
Done.

Gumix commented on src/box/iproto.cc:
```diff
@@ -2660,8 +2728,23 @@ tx_process_override(struct cmsg *m)
 		body_end = body + msg->header.body[0].iov_len;
 	}
 
+	/*
+	 * If we took the `override_route', there must exist either request
+	 * type-specific or unknown request type handler. Their availability
+	 * is checked by the IPROTO thread.
+	 */
```
Yes. If I'm not wrong, `iproto_thread->req_handlers` and `tx_req_handlers` are always in sync, because `iproto_do_cfg()` is implemented via synchronous call over cbus. I'd also ask @CuriousGeorgiy to verify this.

drewdzzz replied:
So, if `iproto` -> `tx` messages are asynchronous, there is a chance a message was sent, but the handlers were deleted. Hence, this comment and an assertion below are not fair.

drewdzzz replied:
Discussed f2f, it's OK.

Gumix commented on src/box/iproto.cc:
```diff
@@ -2631,22 +2638,83 @@ tx_process_replication(struct cmsg *m)
+ * table for this key.
+ */
+static struct iproto_req_handlers *
+iproto_req_handlers_init(uint32_t req_type)
```
Fixed.

Gumix commented on src/box/iproto.cc:
```diff
@@ -2631,22 +2638,83 @@ tx_process_replication(struct cmsg *m)
  */
 static void
-tx_process_override(struct cmsg *m)
+iproto_req_handlers_free(struct iproto_req_handlers *handlers,
```
Fixed.

Gumix commented on src/box/iproto.cc:
```diff
@@ -2631,22 +2638,83 @@ tx_process_replication(struct cmsg *m)
+
+/**
+ * Returns `true' if a handler in `handlers' is set.
+ */
```
Done.

Gumix commented on src/box/iproto.cc:
```diff
@@ -41,6 +41,7 @@
 
```
Done.

CuriousGeorgiy commented on src/box/iproto.cc:
```diff
@@ -3778,39 +3849,30 @@ iproto_override(uint32_t req_type, iproto_handler_t cb,
+
+	bool new_is_set = iproto_req_handler_is_set(handlers);
+	if (!new_is_set && handlers != NULL)
+		iproto_req_handlers_delete(handlers, req_type);
```
Looks like a bug: you can delete a handler from TX only after notifying IPROTO. Otherwise there can be inflight requests that arrive at the override route only to see there is no request handler set (which is unexpected).

Gumix replied:
Thanks, fixed.

CuriousGeorgiy commented on src/box/iproto.cc:
```diff
@@ -3519,6 +3538,48 @@ iproto_override_event_init(struct event *event, void *arg)
+
+	bool new_is_set = iproto_req_handler_is_set(handlers);
+	if (!new_is_set)
+		iproto_req_handlers_delete(handlers, type);
```
Looks like a bug: you can delete a handler from TX only after notifying IPROTO. Otherwise there can be inflight requests that arrive at the override route only to see there is no request handler set (which is unexpected).

CuriousGeorgiy replied:
BTW, please also refactor the common logic out into a separate helper.

Gumix replied:
Thanks, fixed.

> BTW, please also refactor the common logic out into a separate helper.

I tried, but couldn't make the code more beautiful than it is now.

CuriousGeorgiy replied:
I mean these two snippets are very similar, I suggest refactoring them into a separate function:

https://github.com/Gumix/tarantool/blob/f5953480a14118dd06efb8f5668b398afc5f4637/src/box/iproto.cc#L3653-L3660

https://github.com/Gumix/tarantool/blob/f5953480a14118dd06efb8f5668b398afc5f4637/src/box/iproto.cc#L4322-L4329

Feel free to ignore.

Gumix replied:
Moved to the new function `iproto_override_finish()`.

CuriousGeorgiy commented on src/box/lua/iproto.lua:
```diff
@@ -0,0 +1,30 @@
+    if callback ~= nil and type(callback) ~= 'function' then
+        box.error(box.error.PROC_LUA,
+                  ("bad argument #2 to 'override' (function expected, " ..
+                   "got %s)"):format(type(callback)))
```
I am not an expert on the trigger subsystem, but can't it also be a callable table now? @drewdzzz

Gumix replied:
AFAIK, it's hard to support callable tables here. Let's allow only functions here. Callable objects can be set via `trigger.set()`.

drewdzzz replied:
We cannot check if an object is callable from Lua (for example, AFAIK, there is no way to check if `cdata` is callable from Lua). But if we really want to accept callable objects, we could call `trigger.set` with `pcall` or rewrite this function in C. But as for me, it's OK to accept only functions here.

CuriousGeorgiy commented on src/box/lua/iproto.lua:
```diff
@@ -0,0 +1,30 @@
+    if type(request_type) ~= 'number' then
+        box.error(box.error.PROC_LUA,
+                  ("bad argument #1 to 'override' (number expected, " ..
+                   "got %s)"):format(type(callback)))
```
Can't it also be a string now?

Gumix replied:
Done.

CuriousGeorgiy commented on src/box/lua/iproto.lua:
```diff
@@ -0,0 +1,30 @@
+    end
+
+    local event_name = ('box.iproto.override[%d]'):format(request_type)
+    local trigger_name = 'box.iproto.override.lua.user_handler'
```
What if a user incidentally has a trigger with the same name?

Gumix replied:
The new trigger will override the old one. That's why I called it "__tarantool__internal__iproto__override__trigger__" in the first PR iteration.

drewdzzz replied:
In the triggers documentation (which doesn't exist yet) we will write that `tarantool` and `box` root namespaces are reserved for our purposes.

Gumix commented on src/box/iproto.cc:
```diff
@@ -275,15 +275,22 @@ unsigned iproto_readahead = 16320;
-	iproto_handler_destroy_t destroy;
-	/** Context passed to request handler callback and destructor. */
-	void *ctx;
+struct iproto_req_handlers {
```
I don't like `registry`, because it is a single entry in the hash table, not a registry.

Gumix commented on src/box/lua/func_adapter.c:
```diff
@@ -141,10 +141,10 @@ func_adapter_lua_push_null(struct func_adapter_ctx *base)
 static void
 func_adapter_lua_push_msgpack(struct func_adapter_ctx *base, const char *data,
-			      const char *data_end)
+			      const char *data_end, void *mp_ctx)
```
Done.

Gumix commented on src/box/space.c:
```diff
@@ -813,12 +813,12 @@ space_run_replace_triggers(struct event *event, struct txn *txn,
 			assert(row != NULL && row->header != NULL);
 			func_adapter_push_msgpack(trigger, &ctx, row->header,
-						  row->header_end);
+						  row->header_end, NULL);
```
Done.

Gumix commented on src/lib/core/func_adapter.h:
```diff
@@ -81,7 +81,7 @@ struct func_adapter_vtab {
 	 */
 	void (*push_msgpack)(struct func_adapter_ctx *ctx, const char *data,
-			     const char *data_end);
+			     const char *data_end, void *mp_ctx);
```
Done.

Gumix commented on src/lib/core/func_adapter.h:
```diff
@@ -226,9 +226,9 @@ func_adapter_push_null(struct func_adapter *func, struct func_adapter_ctx *ctx)
 func_adapter_push_msgpack(struct func_adapter *func,
 			  struct func_adapter_ctx *ctx,
-			  const char *data, const char *data_end)
+			  const char *data, const char *data_end, void *mp_ctx)
```
Done.

Gumix commented on src/box/iproto.cc:
```diff
@@ -275,15 +275,22 @@ unsigned iproto_readahead = 16320;
- * Request handler meta information.
+ * Request handlers meta information. The IPROTO request of each type can be
+ * overridden by the following types of handlers (listed in priority order):
+ *  1. Lua or C handler, set by `iproto_override()'.
```
Done.

Gumix commented on src/box/iproto.cc:
```diff
@@ -2631,22 +2638,83 @@ tx_process_replication(struct cmsg *m)
+	TRASH(handlers);
+	free(handlers);
+
+	if (req_type != iproto_type_MAX) {
```
Split into `iproto_req_handlers_delete` and `mh_req_handlers_del`.
Same for `iproto_req_handlers_new` and `mh_req_handlers_put`.

Gumix commented on src/box/iproto.cc:
```diff
@@ -2704,7 +2746,88 @@ iproto_req_handler_is_set(struct iproto_req_handlers *handlers)
+			free(type_name_lower);
+			break;
+		}
+		free(type_name_lower);
```
> We will now have 2 places where we do this lowercase conversion (see also src/box/lua/iproto.c).

In `src/box/lua/iproto.c` the `iproto_key_strs` is lowered, not `iproto_type_strs`.

> How about creating an array of lowercase string constants one time during iproto initialization?

Done.

> That way you also use strindex from src/lib/core/utils.c here.

`strindex()` ignores the case of the characters, but I need case-sensitive comparison here, because from the trigger-registry point of view the events "box.iproto.override.ping" and "box.iproto.override.PING" are totally unique. For this reason only lowercase names are allowed in `get_iproto_type_by_name()` to avoid any further inconsistency.

Gumix commented on src/box/iproto.cc:
```diff
@@ -2701,22 +2722,211 @@ tx_process_replication(struct cmsg *m)
+
+	while (event_trigger_iterator_next(&it, &trigger, &name)) {
+		struct mp_ctx mp_ctx_header, mp_ctx_body;
+		mp_ctx_create_default(&mp_ctx_header, iproto_key_translation);
```
I'm not sure about this part. Is it necessary to call `mp_ctx_create_default()` on every iteration?

drewdzzz replied:
Also, can we use one `mp_ctx` for both packets?

CuriousGeorgiy replied:
Yes.

Gumix replied:
Actually it doesn't work, because `func_adapter_push_msgpack_with_ctx` calls `mp_ctx_move`.

drewdzzz commented on src/box/iproto.cc:
```diff
@@ -3377,6 +3510,94 @@ iproto_thread_init(struct iproto_thread *iproto_thread)
+
+	if (type == iproto_type_MAX) {
+		say_crit("The event `%s' is in IPROTO override namespace, but "
+			 "`%s' is not a valid request type", name, req_name);
```
It's not correct to use `req_name` here since it was advanced during the name parsing. Let's use `name + prefix_name`.

Gumix replied:
Changed to:
```diff
@@ -3650,11 +3654,13 @@ get_iproto_type_from_event_name(const char *name, bool *is_by_id)
                return iproto_type_MAX;

        const char *req_name = name + prefix_len;
+       const char *req_name_err = req_name;
        uint32_t type = iproto_type_MAX;

        if (*req_name == '.') {
                /* Skip the dot. */
                req_name++;
+               req_name_err = req_name;
                type = get_iproto_type_by_name(req_name);
        } else if (*req_name == '[') {
                /* Skip open bracket. */
@@ -3674,11 +3680,14 @@ get_iproto_type_from_event_name(const char *name, bool *is_by_id)
                                        type = iproto_type_MAX;
                        }
                }
+       } else {
+               return iproto_type_MAX;
        }

        if (type == iproto_type_MAX) {
                say_crit("The event `%s' is in IPROTO override namespace, but "
-                        "`%s' is not a valid request type", name, req_name);
+                        "`%s' is not a valid request type",
+                        name, req_name_err);
                return iproto_type_MAX;
        }
        bool is_unsupported_request = type == IPROTO_JOIN ||
```

drewdzzz commented on src/box/iproto.cc:
```diff
@@ -2774,6 +2792,30 @@ mh_req_handlers_del(uint32_t req_type)
+	event_ref(event);
+
+	if (is_by_id) {
+		if (handlers->event_by_id != NULL)
```
Why do you need to replace event with another one?

drewdzzz reviewed (commented):
Please address my comments.
Also:
1. Trigger documentation is already written, so you can write doc for this feature.
2. Shouldn't we mark `box.iproto.override` change as breaking?

Gumix commented on src/box/space.c:
```diff
@@ -887,15 +889,21 @@ space_run_replace_triggers(struct event *event, struct txn *txn,
 				       iproto_type_name(stmt->type));
 		/* Pass xrow header and body to recovery triggers. */
 		if (stmt->space->run_recovery_triggers) {
+			struct mp_ctx mp_ctx_header, mp_ctx_body;
+			mp_ctx_create_default(&mp_ctx_header,
+					      iproto_key_translation);
+			mp_ctx_create_default(&mp_ctx_body,
+					      iproto_key_translation);
```
It doesn't work, because `func_adapter_push_msgpack_with_ctx` calls `mp_ctx_move`.

Gumix commented:
> Trigger documentation is already written, so you can write doc for this feature.

Added TarantoolBot request.

> Shouldn't we mark box.iproto.override change as breaking?

Added:
```diff
diff --git a/changelogs/unreleased/gh-8138-switch-box-iproto-override-to-triggers.md b/changelogs/unreleased/gh-8138-switch-box-iproto-override-to-triggers.md
new file mode 100644
index 000000000..8c1fc1d1e
--- /dev/null
+++ b/changelogs/unreleased/gh-8138-switch-box-iproto-override-to-triggers.md
@@ -0,0 +1,5 @@
+## feature/core
+
+* `box.iproto.override()` was switched to the universal trigger registry. As a
+  side effect, now this function does not raise an error when a wrong request
+  type is passed. All such errors are logged with `CRITICAL` level (gh-8138).
```

drewdzzz commented on src/box/iproto.cc:
```diff
@@ -3482,6 +3619,122 @@ iproto_thread_init(struct iproto_thread *iproto_thread)
+	const char *prefix = "box.iproto.override";
+	const size_t prefix_len = strlen(prefix);
+	if (strncmp(name, prefix, prefix_len) != 0)
+		return iproto_type_MAX;
```
I'm not sure it's a good decision because in future we can have a request type with value 129 (current `iproto_type_MAX` value).
I think we shouldn't have any reserved values here - let's return bool indicating if event name was successfully parsed and return value with an output argument.
Also, let's test this case if we can.

Gumix replied:
Done.

> Also, let's test this case if we can.

Added a testcase with hardcoded value 129.

drewdzzz reviewed (commented):
Sorry for such an "iterative" review.
Generally, it's OK, just one last detail in the comment above.
